### DISCUSSION TITLE
What's the best form input field for floats?

### DISCUSSION THREAD

**Discussion by ledell** [h2oai/wave#1827] [Q&A] [answered]
I can't find a form input field that looks like it's good for floats.  The [spinbox](https://wave.h2o.ai/docs/widgets/form/spinbox) seems useful for integers, but the only thing I can find for float input would be the [textbox](https://wave.h2o.ai/docs/widgets/form/textbox), and then convert to a float on the backend.  Just wanted to check if there's a better built-in solution available.

If there's not something available, maybe we could make a new field type called floatbox which is a wrapper for a textbox that verifies that the input is a float.

**Comment by mturoci** [comment, marked as answer]
Hi Erin, try setting the `step` attribute to a float value (e.g. step=0.1) when using spinbox. This should allow for floats with the precision you need. Defaults to 1.

**Reply by DeeJaayArgh** [reply]
I thinking spinbox could become even more useful for floats if the spin functionality could be disabled (no up/down icons and disable keyboard callbacks for up arrow down arrow), so it just allows entry of numbers.  Maybe this is already possible using masks with ui.textbox.  For example if I have a field that can be any value from 0 to 100 but want to allow entry of various numbers with variable precision  (50, 3.14, 5.1234354), setting step to something small like 0.0000001 is awkward.

**Reply by mturoci** [reply]
I see what you are trying to say, but it's not a big deal IMO as the up/down arrows are optional and the user is free to type whatever float she wants. You can think of `step` as a way to define precision and optional arrow step. We messed up on naming here so I will need to improve the docs on this one.

**Comment by ledell** [comment]
@mturoci Adjusting the `step` size to a float in the spinbox, allows float input now. (for some reason I assumed it was int only.. oops) Thank you!  
 
A question I have though is -- it only allows values of the step increments.  So the spinner goes from 0.0 to 1.0 and step is 0.1, the user cannot type in something like 0.25 (at least it does not seem like it when I tried).  So I guess then maybe what I really need is the textbox with float validation because I am using Wave for a UI for inputting ML hyper-parameters, which often have bounds of 0.0 and 1.0, but can be any value inbetween.  What do you think?

**Reply by mturoci** [reply]
> So the spinner goes from 0.0 to 1.0 and step is 0.1, the user cannot type in something like 0.25 

I assume you used step as 0.1? Depends on the precision you want to allow. if 2 decimal places, then the step should be 0.01. If 3, then 0.001. Together with min=0.1 and max=1 this should allow for:

* Users can type up to your desired precision or less meaning step 0.01 will allow both 0.2 and 0.25.
* Configure up/down buttons to go by the given step - most people will most likely just write the float by hand though.

> maybe what I really need is the textbox with float validation

This is an option as well, but it's the harder way and you will basically achieve spinbox without up/down arrows for much more effort.

**Reply by mturoci** [reply]
However, the way you and @DeeJaayArgh assumed it to work - step is independent of the precision seems the most natural so I think we should introduce a new attr called `max_precision` that would separate these two. Wdyt?

**Reply by ledell** [reply]
Yes I think that would be super helpful, thanks @mturoci!

**Reply by DeeJaayArgh** [reply]
I went with ui.textbox since I only need floating point validation and not the step functionality. I did something like the following  to enter floating point values.  You could easily add additional checking for min and max too.  This doesn't prevent user from typing invalid inputs, it just alerts them when input is invalid

```
error_msg = None

# This conditional gets hit every time the textbox contents change (I think something like 0.5s delay on it to not interrupt user typing)
if q.args.float_field:
    value = q.args.float_field
    try:
        value = float(value)
    except ValueError:
        error_msg= "Not a valid floating point number"
else:
    value = 3.1415

# Use the follwing textbox in your form
ui.textbox(
    name="float_field", 
    value=value,
    trigger=True, 
    error=error_msg)
```
